modules/phase_field/src/kernels/CHInterface.C

```diff
@@ -13,54 +13,97 @@ InputParameters validParams<CHInterface>()
   params.addClassDescription("Gradient energy Cahn-Hilliard Kernel");
   params.addRequiredParam<std::string>("kappa_name", "The kappa used with the kernel");
   params.addRequiredParam<std::string>("mob_name", "The mobility used with the kernel");
-  params.addParam<std::string>("Dmob_name", "DM", "The D mobility used with the kernel");
-  params.addRequiredParam<std::string>("grad_mob_name", "The gradient of the mobility used with the kernel");
-  params.addParam<bool>("has_MJac", false, "Jacobian information for the mobility is defined");
+  params.addCoupledVar("args", "Vector of arguments to mobility");
 
   return params;
 }
 
 CHInterface::CHInterface(const std::string & name, InputParameters parameters) :
-    Kernel(name, parameters),
+    DerivativeMaterialInterface<JvarMapInterface<Kernel> >(name, parameters),
     _kappa_name(getParam<std::string>("kappa_name")),
     _mob_name(getParam<std::string>("mob_name")),
-    _Dmob_name(getParam<std::string>("Dmob_name")),
-    _grad_mob_name(getParam<std::string>("grad_mob_name")),
     _kappa(getMaterialProperty<Real>(_kappa_name)),
     _M(getMaterialProperty<Real>(_mob_name)),
-    _has_MJac(getParam<bool>("has_MJac")),
-    _DM(_has_MJac ? &getMaterialProperty<Real>(_Dmob_name) : NULL),
-    _grad_M(getMaterialProperty<RealGradient>(_grad_mob_name)),
-    _Dgrad_Mnp(_has_MJac ? &getMaterialProperty<RealGradient>("Dgrad_Mnp") : NULL),
-    _Dgrad_Mngp(_has_MJac ? &getMaterialProperty<Real>("Dgrad_Mngp") : NULL),
+    _dMdc(getMaterialPropertyDerivative<Real>(_mob_name, _var.name())),
+    _d2Mdc2(getMaterialPropertyDerivative<Real>(_mob_name, _var.name(), _var.name())),
     _second_u(second()),
     _second_test(secondTest()),
-    _second_phi(secondPhi())
+    _second_phi(secondPhi()),
+    _nvar(_coupled_moose_vars.size()),
+    _dMdarg(_nvar),
+    _d2Mdcdarg(_nvar),
+    _d2Mdargdarg(_nvar),
+    _coupled_grad_vars(_nvar)
 {
+  // Iterate over all coupled variables
+  for (unsigned int i = 0; i < _nvar; ++i)
+  {
+    //Set material property values
+    _dMdarg[i] = &getMaterialPropertyDerivative<Real>(_mob_name, _coupled_moose_vars[i]->name());
+    _d2Mdcdarg[i] = &getMaterialPropertyDerivative<Real>(_mob_name, _var.name(), _coupled_moose_vars[i]->name());
+    _d2Mdargdarg[i].resize(_nvar);
+    for (unsigned int j = 0; j < _nvar; ++j)
+      _d2Mdargdarg[i][j] = &getMaterialPropertyDerivative<Real>(_mob_name, _coupled_moose_vars[i]->name(), _coupled_moose_vars[j]->name());
+
+    //Set coupled variable gradients
+    _coupled_grad_vars[i] = &coupledGradient("args", i);
+  }
 }
 
 Real
 CHInterface::computeQpResidual()
 {
-  return _kappa[_qp] * _second_u[_qp].tr() * (_M[_qp] * _second_test[_i][_qp].tr() + _grad_M[_qp] * _grad_test[_i][_qp]);
+  RealGradient grad_M = _dMdc[_qp]*_grad_u[_qp];
```

Comment:
If the standard is to always leave a space around the operators, that is OK. But I much prefer it the way that I described above.